core/src/main/java/org/eclipse/hono/util/AddressHelper.java

```diff
@@ -0,0 +1,82 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.util;
+
+import java.util.Objects;
+import java.util.regex.PatternSyntaxException;
+
+/**
+ * Utility class for handling AMQP addresses.
+ */
+public class AddressHelper {
+
+    private AddressHelper() {
+    }
+
+    /**
+     * Gets the AMQP <em>target</em> address to use for creating with specific Hono endpoint.
+     *
+     * @param endpoint The endpoint for the address (e.g. telemetry).
+     * @param tenantId The tenant of the address.
+     * @param resourceId The device to upload data for. If {@code null}, the target address can be used
+     *                 to upload data for arbitrary devices belonging to the tenant.
+     * @param addressRewriteRule The rewrite rule to be applied to the address.
+     *                           See {@link #rewrite(String, String)} for more information about syntax and behavior of this property.
+     * @return The target address.
+     * @throws NullPointerException if endpoint or tenant is {@code null}.
+     */
+    public static String getTargetAddress(final String endpoint, final String tenantId, final String resourceId, final String addressRewriteRule) {
+        final StringBuilder addressBuilder = new StringBuilder(Objects.requireNonNull(endpoint))
+                .append("/").append(Objects.requireNonNull(tenantId));
+        if (!Strings.isNullOrEmpty(resourceId)) {
+            addressBuilder.append("/").append(resourceId);
+        }
+
+        return AddressHelper.rewrite(addressBuilder.toString(), addressRewriteRule);
+    }
+
+    /**
+     * Rewrites address for a certain endpoint according to the provided rule.
+     *
+     * @param address The address to be rewritten.
+     * @param rewriteRule The rewrite rule. The rule is in the <em>$PATTERN $SUBSTITUTION</em> format.
+     *                    Pattern and substitution use the regular Java pattern syntax.
+     *                    The pattern should match the original address.
+     *
+     *                    Example:
+     *                    <em>([a-z_]+)/([\\w-]+) test-vhost/$1/$2</em>
+     *                    will rewrite
+     *                    <em>telemetry/DEFAULT_TENANT</em> to <em>test-vhost/telemetry/DEFAULT-TENANT</em>
+     *
+     *                    If the rule is {@code null}, empty, in the wrong format or the pattern doesn't match the address,
+     *                    the original address will be returned.
+     * @return The address to be used.
+     */
+    public static String rewrite(final String address, final String rewriteRule) {
+        if (Strings.isNullOrEmpty(rewriteRule)) {
+            return address;
+        }
+
+        final String[] elements = rewriteRule.split(" ", 2);
+        if (elements.length != 2) {
+            return address;
+        }
+
+        try {
+            return address.replaceAll(elements[0], elements[1]);
```

Comment:
I still think we should use pre-compiled patterns for this.

Reply:
It makes sense if we initialize AddressHelper and reuse that Pattern instead of using it statically. I'll take a look at that.